mods/default/nodes.lua

```diff
@@ -1614,6 +1617,90 @@ minetest.register_node("default:bush_sapling", {
 	end,
 })
 
+minetest.register_node("default:blueberry_bush_leaves_with_berries", {
+	description = "Blueberry Bush Leaves with Berries",
+	drawtype = "allfaces_optional",
+	waving = 1,
+	tiles = {"default_blueberry_bush_leaves.png^default_blueberry_overlay.png"},
+	paramtype = "light",
+	groups = {snappy = 3, flammable = 2, leaves = 1, dig_immediate = 3},
+	drop = "default:blueberries",
+	sounds = default.node_sound_leaves_defaults(),
+	node_dig_prediction = "default:blueberry_bush_leaves",
+
+	after_dig_node = function(pos, oldnode, oldmetadata, digger)
+		minetest.set_node(pos, {name = "default:blueberry_bush_leaves"})
+		minetest.get_node_timer(pos):start(math.random(300, 1500))
+	end,
+})
+
+minetest.register_node("default:blueberry_bush_leaves", {
+	description = "Blueberry Bush Leaves",
+	drawtype = "allfaces_optional",
+	waving = 1,
+	tiles = {"default_blueberry_bush_leaves.png"},
+	paramtype = "light",
+	groups = {snappy = 3, flammable = 2, leaves = 1},
+	drop = {
+		max_items = 1,
+		items = {
+			{items = {"default:blueberry_bush_sapling"}, rarity = 5},
+			{items = {"default:blueberry_bush_leaves"}}
+		}
+	},
+	sounds = default.node_sound_leaves_defaults(),
+
+	on_timer = function(pos, elapsed)
+		local nodes = minetest.find_nodes_in_area(vector.add(pos, {x = -1, y = -1, z = -1}),
+			vector.add(pos, {x = 1, y = -1, z = 1}),
+			{"default:dirt_with_grass", "default:dirt_with_snow"})[1]
```

Comment:
This means if grass is removed, berries stop growing, then if grass grows back, they still don't grow, and that could be confusing for players. There's also no real-world parallel here. Berries don't only grow if there's grass nearby. Can we remove this check?

Reply:
If you decide to have a check here it should be for 'soil' group.

Reply:
I considered suggesting that, but I think it's better to avoid the check.

Reply:
I considered making the leaf nodes attached like grass, but then some bushes which had grown on hills would be "unstable". I vote for a soil group check, because it seems very strange that players could have blueberry plants producing blueberries without the presence of any dirt whatsoever.

Reply:
I agree with removing the check. A soil node could be temporarily removed from under and the timer trigger during that time, this would create leaves that never grow berries again.
Removing a soil node from under doesn't kill the leaves, so growing berries in that situation seems ok.
Also a small advantage in avoiding node searches.